### PR TITLE
PR for code review

### PR DESCRIPTION
**Pull request by elezar**

### CODE REVIEW

**Review by elezar — COMMENTED** [review]
Thanks for the suggestions. I have made some initial comments below.

File: deployments/gpu-operator/labels.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package main
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/labels.go:1]: done

File: deployments/gpu-operator/labels.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+func MaybeSetPaused(value string) string {
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/labels.go:3]: done

File: deployments/gpu-operator/main.go [hunk 3]
```diff
@@ -39,7 +39,7 @@ const (
 	ResourceNodes  = "nodes"
 	MigConfigLabel = "nvidia.com/mig.config"
 
-	DefaultReconfigureScript         = "/usr/bin/reconfigure-mig.sh"
+	DefaultReconfigureScript         = "/usr/bin/reconfigure-mig.go"
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/main.go:42]: changing back to reconfigure-mig.sh

File: deployments/gpu-operator/main.go [hunk 4]
```diff
@@ -325,10 +325,11 @@ func runScript(migConfigValue string) error {
 	if withShutdownHostGPUClientsFlag {
 		args = append(args, "-d")
 	}
-	cmd := exec.Command(reconfigureScriptFlag, args...)
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	return cmd.Run()
+	hostGPUClientServices := strings.Join(gpuClients.SystemdServices, ",")
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/main.go:328]: I was doing the same i.e. passing slice as an argument. But was getting error. Let me try again.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 5]
```diff
@@ -0,0 +1,982 @@
...
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+var (
+	hostGPUClientServicesStopped = []string{}
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:29]: these variables are used in setState function() and their values are only assigned in shuttingGPUClients() and restartingGPUClients(). However, setState() function is called in multiple places throughout the script even before reaching restartingGPUClients().

File: deployments/gpu-operator/reconfigure-mig.go [hunk 6]
```diff
@@ -0,0 +1,982 @@
...
+	if [ "${?}" != "0" ]; then
+		exit_failed
+	fi
+*/
+
+func applyMigConfig(nvidiaMigPartedAlias string, migConfigFile string, selectedMIGConfig string, currentLabels map[string]string, hostGPUClientServices []string, withShutdownHostGPUClients bool, hostRootMount string, nodeName string) error {
+	fmt.Println("Applying the selected MIG config to the node")
+	cmd := exec.Command(nvidiaMigPartedAlias, "-d", "apply", "-f", migConfigFile, "-c", selectedMIGConfig)
+	if err := cmd.Run(); err != nil {
+		setStateAndExit("failed", 1, currentLabels, hostGPUClientServices, withShutdownHostGPUClients, hostRootMount, nodeName)
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:776]: moved the setState() function to the reconfigure() function. But, that didn't help in reducing the number of parameters. Since setState() function need all these parameters in isolation itself.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 7]
```diff
@@ -0,0 +1,982 @@
...
+	}
+	return nil
+}
+
+/*
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:781]: done

File: deployments/gpu-operator/reconfigure-mig.go [hunk 8]
```diff
@@ -0,0 +1,982 @@
...
+	}
+	return nil
+}
+
+func hostStartSystemdServices(hostRootMount string, hostGPUClientServices []string) error {
+	// If HOST_GPU_CLIENT_SERVICES_STOPPED is empty, then it's possible that
+	// host_stop_systemd_services was never called, so let's double check to see
+	// if there's anything we should actually restart.
+	if len(hostGPUClientServicesStopped) == 0 {
+		for _, s := range hostGPUClientServices {
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:942]: created a new function hostStopSystemdServicesInactive() for all the remaining cases apart from "is-active" and used that function here as well.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 9]
```diff
@@ -48,9 +58,12 @@ func reconfigure(nodeName string, migConfigFile string, selectedMIGConfig string
 		return err
 	}
 
+	// Declaring and initializing a struct using a struct literal
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:61]: Is this comment useful?

File: deployments/gpu-operator/reconfigure-mig.go [hunk 10]
```diff
@@ -48,9 +58,12 @@ func reconfigure(nodeName string, migConfigFile string, selectedMIGConfig string
...
+	migConfigProperties := migConfig{nvidiaMigPartedAlias, migConfigFile}
+
 	/*
 		retrieves current values of kubernetes node labels
-		i/p: nodeName ,(device-plugin, gpu-feature-discovery, dcgm-exporter, dcgm, nvsm)
+		i/p: nodeName, (device-plugin, gpu-feature-discovery, dcgm-exporter, dcgm, nvsm)
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:66]: Comment is out of date.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 11]
```diff
@@ -63,8 +76,8 @@ func reconfigure(nodeName string, migConfigFile string, selectedMIGConfig string
 		i/p: migConfigFile, selectedMigConfig
 		o/p: true or false
 	*/
-	if err := assertConfigPresent(nvidiaMigPartedAlias, migConfigFile, selectedMIGConfig); err != nil {
-		setStateAndExit("failed", 1, currentLabels, hostGPUClientServices, withShutdownHostGPUClients, hostRootMount, nodeName)
+	if err := assertConfigPresent(migConfigProperties, selectedMIGConfig); err != nil {
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:79]: For all functions where `migConfigProperties` is now the first argument, we could consider hanging these functions off the type instead. For example:
```
func (c migConfig) assertConfigPresent(config string) error {
}
```
and then call it using `migConfigProperties.assertConfigPresent(selectedMIGConfig)` instead.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 12]
```diff
@@ -75,40 +88,41 @@ func reconfigure(nodeName string, migConfigFile string, selectedMIGConfig string
...
 
 	/*
-		host_persist_config
+		TODO: rightConfigApplied
 		ensuring that right config is applied to the node if the MIG Parted is also installed as a systemd service
-		i/p: hostRootMount, hostMIGManagerStateFile, selectedMIGConfig
+		i/p: hostRootMount, hostMIGManagerStateFile, selectedMIGConfig, currentLabels, hostGPUClientServices, nodeName, withShutdownHostGPUClients
 		o/p: true or false
+
+		if err := rightConfigApplied(hostRootMount, hostMIGManagerStateFile, selectedMIGConfig, currentLabels, hostGPUClientServices, nodeName, withShutdownHostGPUClients); err != nil{
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:111]: I think this function name is misleading. It does not properly reflect what we want to achieve here.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 13]
```diff
@@ -661,97 +481,40 @@ func hostStopSystemdServices(hostRootMount string, hostGPUClientServices []strin
...
-
-fi
-*/
-func applyMIGModeChange(nvidiaMigPartedAlias string, migConfigFile string, selectedMIGConfig string,
-	 nodeName string, withReboot bool, hostRootMount string, currentLabels map[string]string, hostGPUClientServices []string, withShutdownHostGPUClients bool) error {
-	applyCmd := exec.Command(nvidiaMigPartedAlias, "-d", "apply", "--mode-only", "-f", migConfigFile, "-c", selectedMIGConfig)
-	assertCmd := exec.Command(nvidiaMigPartedAlias, "-d", "assert", "--mode-only", "-f", migConfigFile, "-c", selectedMIGConfig)
+// Applying the MIG mode change from the selected config to the node (and double checking it took effect)
+func applyMIGModeChange(migConfigProperties migConfig, selectedMIGConfig string,
+	nodeName string, withReboot bool, hostRootMount string, currentLabels map[string]string, hostGPUClientServices []string, withShutdownHostGPUClients bool) error {
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:491]: Instead of passing all the parameters required to set the sate here, let's return a `rebootRequired` boolean value instead and respond to this in the `reconfigure` function. This will further clean up where we call `setState`.

So we would have:
```suggestion
func applyMIGModeChange(migConfigProperties migConfig, selectedMIGConfig string) (bool, error) {
```

File: deployments/gpu-operator/reconfigure-mig.go [hunk 14]
```diff
@@ -661,97 +481,40 @@ func hostStopSystemdServices(hostRootMount string, hostGPUClientServices []strin
...
-	if err := assertCmd.Run(); err != nil {
+	if err := migModeSettingApplied(migConfigProperties, selectedMIGConfig); err != nil {
 		fmt.Println("Failed to assert MIG mode change: ", err)
 		if withReboot {
 			rebootNode(nodeName, hostRootMount, currentLabels, hostGPUClientServices, withShutdownHostGPUClients)
 		}
 		return err
 	}
-	fmt.Println("MIG mode change applied successfully")
+	klog.Info("MIG mode change applied successfully")
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:505]: Let's either switch all `fmt.Print*` statements to `klog` or none.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 15]
```diff
@@ -167,21 +177,21 @@ func reconfigure(nodeName string, migConfigFile string, selectedMIGConfig string
...
 		return err
 	}
 
-	setStateAndExit("success", 0, currentLabels, hostGPUClientServices, withShutdownHostGPUClients, hostRootMount, nodeName)
+	setState("success", 0, currentLabels, hostGPUClientServices, withShutdownHostGPUClients, hostRootMount, nodeName)
 	return nil
 }
 
-func setStateAndExit(state string, exitCode int, currentLabels map[string]string, hostGPUClientServices []string, withShutdownHostGPUClients bool, hostRootMount string, nodeName string) {
+func setState(state string, exitCode int, currentLabels map[string]string, hostGPUClientServices []string, withShutdownHostGPUClients bool, hostRootMount string, nodeName string) {
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:188]: What may make this function easier to reason about and use is to have it accept an `error` instead of an `exitCode` and also return an `error` This would mean that we would use it as: `return setState("failed", err, ...)` or `return setState("success", nil, ...)` in our calling function.

Also note that this is still running `os.Exit()` at the end which is not our desired flow.

File: deployments/gpu-operator/reconfigure-mig.go [hunk 16]
```diff
@@ -0,0 +1,603 @@
...
+		cmd := exec.Command("kubectl", args...)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		if err := cmd.Run(); err != nil {
+			fmt.Println("Unable to bring up GPU client pods by setting their daemonset labels")
+			exitCode = 1
+		}
+	}
+
+	fmt.Printf("Changing the 'nvidia.com/mig.config.state' node label to '%s'\n", state)
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:220]: Question: Why don't we call `setStateToX` here?

File: deployments/gpu-operator/reconfigure-mig.go [hunk 17]
```diff
@@ -0,0 +1,603 @@
...
+	if err != nil {
+		fmt.Printf("Unable to set 'nvidia.com/mig.config.state' to '%s'\n", state)
+		fmt.Println("Exiting with incorrect value in 'nvidia.com/mig.config.state'")
+		exitCode = 1
+	}
+	os.Exit(exitCode)
+}
+
+// initMigParted sets up mig-parted on the host if required and sets the config file.
+func initMigParted(hostRootMount string, hostNvidiaDir string, withShutdownHostGpuClinets bool, migConfigFile string) (string, string, error) {
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:232]: Do we ever return an `error` here?

File: deployments/gpu-operator/reconfigure-mig.go [hunk 18]
```diff
@@ -0,0 +1,603 @@
...
+	}
+	os.MkdirAll(fmt.Sprintf("%s/%s/mig-manager/", hostRootMount, hostNvidiaDir), 0755)
+	exec.Command("cp ${which nvidia-mig-parted}", fmt.Sprintf("%s/%s/mig-manager/nvidia-mig-parted", hostRootMount, hostNvidiaDir)).Run()
+	exec.Command("cp", migConfigFile, fmt.Sprintf("%s/%s/mig-manager/config.yaml", hostRootMount, hostNvidiaDir)).Run()
+	nvidiaMigPartedAlias := fmt.Sprintf("chroot %s %s/mig-manager/nvidia-mig-parted", hostRootMount, hostNvidiaDir)
+	migConfigFile = fmt.Sprintf("%s/mig-manager/config.yaml", hostNvidiaDir)
+	return nvidiaMigPartedAlias, migConfigFile, nil
+}
+
+// Getting current value of the 'nvidia.com/gpu.deploy.__' node label,
```
Comment (elezar) [inline deployments/gpu-operator/reconfigure-mig.go:244]: This docstring is not associated with the right function.

File: deployments/gpu-operator/main.go [hunk 19]
```diff
@@ -19,7 +19,7 @@ package main
 import (
 	"fmt"
 	"os"
-	"os/exec"
+	// "os/exec"
```
[the remark this answers is not in the document]
Comment (ayush550) [reply deployments/gpu-operator/main.go:22]: done

File: deployments/gpu-operator/reconfigure-mig.go (hunk 9 above)
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:61]: It was for my reference. 
removed

File: deployments/gpu-operator/reconfigure-mig.go (hunk 10 above)
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:66]: updated

File: deployments/gpu-operator/reconfigure-mig.go (hunk 11 above)
Comment (ayush550) [reply deployments/gpu-operator/reconfigure-mig.go:79]: done

**Review by ayush550 — COMMENTED** [review]
Done with changes.
